enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,491 @@
+}
+```
+
+```go
+// RouteSetOrDeleteHTTPHeaderUnion specifies fields to mention the header details and the action you want to perform on them.
+// +union
+type RouteSetOrDeleteHTTPHeaderUnion struct {
+	// action defines the type of the action to be applied on the header.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:=Set;Delete
+	// +kubebuilder:validation:Required
+	Action RouteHTTPHeaderAction `json:"action,omitempty" protobuf:"bytes,11,opt,name=action,casttype=RouteSetHTTPHeaderAction"`
+
+	// set defines the HTTP header that should be set: added if doesn't exist or replaced if does.
+	// If this field is empty, no header is added or replaced.
+	//
+	// Note that this option only applies to cleartext HTTP connections
+	// and to secure HTTP connections for which the ingress controller
+	// terminates encryption (that is, edge-terminated or reencrypt
+	// connections).  Headers cannot be set for TLS passthrough
+	// connections.
+	//
+	// +kubebuilder:validation:Optional
+	// +unionMember
+	Set *RouteSetHTTPHeader `json:"set,omitempty" protobuf:"bytes,9,opt,name=set"`
+}
+```
+
+```go
+// RouteHTTPHeaderAction defines operation you can perform on headers.
+type RouteHTTPHeaderAction string
+
+const (
+    // Set is used specify set action per header
+    Set    RouteHTTPHeaderAction = "Set"
+    // Delete is used specify delete action per header
+	Delete RouteHTTPHeaderAction = "Delete"
+)
+```
+
+The RouteSetHTTPHeader type has fields for specifying a header name, header value and action.
+```go
+// RouteSetHTTPHeader specifies what value needs to be set on an HTTP header.
+type RouteSetHTTPHeader struct {
+    // Value specifies a header value.
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=4096
+    Value string `json:"value" protobuf:"bytes,2,opt,name=value"`
+}
+```
+
+### Implementation Details/Notes/Constraints
+A) Setting HTTP headers by cluster admin via IngressController's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variable, `ROUTER_HTTP_HEADERS`, which specifies the headers to set or deleted in the
+following frontend sections of haproxy.config: `public`, `fe_no_sni`,`fe_sni`.
+The ingress operator sets this variable based on the IngressController's
+`spec.httpHeaders.headersManipulation.response[*].set` or `spec.httpHeaders.headersManipulation.response[*].delete` field values. 
+
+The value of `ROUTER_HTTP_HEADERS` is a comma-delimited list of percent encoded  triplets of
+the form `name:value:action` for action `Set` and `name:action` pairs for action `Delete`.
+The template translates these values into the appropriate `http-response set-header` stanzas.
```

Comment:
What do we do with formats?  For example, what do we do with the following (inspired by <http://cbonte.github.io/haproxy-dconv/2.2/configuration.html#7.3.2-date>)?
```yaml
spec:
  httpHeaders:
    headersManipulation:
      response: 
      - name: expires
        action: Set
        set:
          value: "%[date(3600),http_date]"
```
Do we allow the format through, or do we we escape it?

Reply:
it will work as percent encoding will encode it %25%5Bdate%283600%29%2Chttp_date%5D
%[date(3600),http_date]
and again decode it %[date(3600),http_date]

Reply:
All right.  We might be able to reuse the `IngressControllerHTTPUniqueIdHeaderPolicy` validation: https://github.com/openshift/api/blob/a7b1b96510144464f4b1123b7750894e0a2d95b7/operator/v1/types_ingress.go#L1268-L1280

We'll probably need to do some additional validation in the operator and router (e.g. verify that the format doesn't contain an invalid converter).

Reply:
For setting dynamic header values for http response. Allowed sample fetchers are res.hdr, ssl_c_der. Allowed converters are lower, base64.
  For setting dynamic header values for http request. Allowed sample fetchers are req.hdr, ssl_c_der. Allowed converters are lower, base64.